### PR TITLE
Fixed XSS

### PR DESCRIPTION
**Pull request by Aravindha1234u**
###### Explanation About What Code Achieves:

  - Bountry URL : https://www.huntr.dev/bounties/2-npm-frappe-charts
  - frappe-charts is vulnerable to Cross-Site Scripting (XSS) due improper data validation
  - I have made a fix in axis chart to validate for template injections and xss.

###### Screenshots/GIFs:

Exploit
image

Fix
image

image

###### Steps To Test:

  - Open the following codesandbox https://codesandbox.io/s/frappe-charts-demo-forked-40w0f?file=/src/index.js
  - Use the payload `""` and place it in `values: [25, ""`
  - XSS payload will get executed.

### CODE REVIEW

File: example.html [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<html>
```
Comment (scmmishra) [inline example.html:1]: Can you remove this file?

File: src/js/utils/axis-chart-utils.js [hunk 2]
```diff
@@ -33,6 +33,19 @@ export function dataPrep(data, type) {
...
+			for(var i=0;i<datasetLength;i++){
+				var temp = data.datasets[i]['values'];
+				for(var j=0;j<temp.length;j++){
+					if(typeof(temp[j]) == "string"){
+						data.datasets[i]['values'][j] = String(temp[j]).replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;");
+					}
+				}
+			}
+		}catch(e){
+		}
```
Comment (scmmishra) [inline src/js/utils/axis-chart-utils.js:47]: A blank `catch` is not desired.

File: src/js/utils/axis-chart-utils.js [hunk 3]
```diff
@@ -33,6 +33,19 @@ export function dataPrep(data, type) {
...
 
+		//XSS
+		try{
+			for(var i=0;i<datasetLength;i++){
+				var temp = data.datasets[i]['values'];
+				for(var j=0;j<temp.length;j++){
+					if(typeof(temp[j]) == "string"){
+						data.datasets[i]['values'][j] = String(temp[j]).replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;");
+					}
+				}
```
Comment (scmmishra) [inline src/js/utils/axis-chart-utils.js:44]: The loop is really inefficient, let's just check if the item is a valid number or not, there's a utility for that

**Review by scmmishra — CHANGES_REQUESTED** [review]
Hey thanks for the PR, but the code quality is not up to the mark. The indentation, variable declarations can be better

**Review by scmmishra — APPROVED** [review]
There's a [line in `axis-chart-utils`](https://github.com/frappe/charts/blob/master/src/js/utils/axis-chart-utils.js#L26) that does check if the value in the dataset is actually a number or not. The fix can be implemented there instead of looping again

`vals = vals.map(val => (!isNaN(val) ? val : 0));`

**Review by scmmishra — CHANGES_REQUESTED** [review]
Address the comments and let me know

File: src/js/utils/axis-chart-utils.js (hunk 3 above)
Comment (Aravindha1234u) [reply src/js/utils/axis-chart-utils.js:44]: Yes it check whether its number or not, But actual problem is not does not work properly. even script are inserted into charts and that leads to XSS

**Comment by Aravindha1234u** [comment]
> There's a [line in `axis-chart-utils`](https://github.com/frappe/charts/blob/master/src/js/utils/axis-chart-utils.js#L26) that does check if the value in the dataset is actually a number or not. The fix can be implemented there instead of looping again
> 
> `vals = vals.map(val => (!isNaN(val) ? val : 0));`

Got It

**Comment by scmmishra** [comment]
> but changing this **vals** does not reflect in the actual dataset that's the problem of this entirely. That the only reason why I created a separate loop for updating the actual dataset;

Patching the loop will be a better option

File: src/js/utils/axis-chart-utils.js (hunk 3 above)
Comment (scmmishra) [reply src/js/utils/axis-chart-utils.js:44]: > even Strings are inserted into charts

The values should always be numbers, strings shouldn't be allowed. For labels a fix for this was already made [here](https://github.com/frappe/charts/pull/315)

**Review by scmmishra — APPROVED** [review]

**Comment by scmmishra** [comment]
Thanks @Aravindha1234u for the fix!

**Comment by adam-nygate** [comment]
@JamieSlome can we track this fix bounty on behalf of 418sec/charts#2?

**Comment by JamieSlome** [comment]
@adam-nygate - the disclosure bounty and fix bounty have both been rewarded here.

Great work all! 🍰

**Pull request merged**